### ISSUE TITLE
Computational graph is not computed ahead of time

### ISSUE THREAD

**Issue by stsievert** [stsievert/adadamp#13]
Currently, the computation graph is not computed ahead of time. That is, the equivalent of this code is being performs:

``` python
import torch

Model = torch.nn.Module

def update(model: Model, grads: Dict[str, Tensor]) -> Model:
    return new_model

def gradient(model: Model) -> Dict[str, Tensor]:
    return grads

model = ...
for update in range(1000):  # for one epoch
    grads = client.compute(client.submit(gradient, ...))
    model = update(model, grads)
```

It'd be better to compute the graph ahead of time, something like this code:

``` python
model_future = client.scatter(model)

for update in range(1000):  # for one epoch
    grads_future = client.submit(gradient, ...)
    model_future = client.submit(model_future, grad_future)

model = client.compute(model_future)
```

I don't this would result in a performance benefit; it'd definitely result in a user-interface benefit.

**Comment by stsievert** [comment]
This might be able to avoid the `deepcopy` line before tasking the gradient computation to different workers, even though mutating inputs is explicitly warned about at https://docs.dask.org/en/latest/delayed-best-practices.html#don-t-mutate-inputs. Here's the use case I'm thinking of:

``` python
from dask.distributed import Client
import dask.array as da
import numpy as np

def update(model, grad):
    assert isinstance(model, np.ndarray) and isinstance(grad, int)
    model -= grad
    return model

if __name__ == "__main__":
    client = Client()
    model = client.scatter(np.zeros(10))
    print(model)  # <Future: finished, type: numpy.ndarray, ...>
    for mul in range(10):
        grad = client.submit(lambda: mul)
        model = client.submit(update, model, grad)
    print(model.result())  # [-45, ...]
    M = client.compute(model)
    print(M.result()) # [-45, ...]
```
